In [1]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
from statsmodels.api import add_constant
from statsmodels.sandbox.regression.gmm import IV2SLS

%store -r perform_dist_high_compare_b3

## Overall

In [2]:
df = pd.read_csv('perform_dist_high_compare_b3.csv', header=0)
df = df.dropna()
df

Unnamed: 0             Student  group  abs_perform_diff_best  phase  \
0             0           lemartinp      3                 270.84      1   
1             1           juligarji      3                 167.50      1   
2             2           nfmorenog      3                 225.00      1   
3             3           AFelipeGA      3                 291.67      1   
4             4          mballeng91      3                 329.17      1   
..          ...                 ...    ...                    ...    ...   
159         159          NicolasPrr      3                 237.50      4   
160         160       jumcorredorro      3                  53.33      4   
161         161          feartheGru      3                 358.33      4   
162         162        Danielsv9207      3                 358.33      4   
163         163  JhonEmmanuelTorres      3                 358.33      4   

     Q7_Q7_1  Q7_Q7_2  Q8_Q8_1  Q10  similarity  
0        0.0      0.0      4.0  1.0    0.146341  
1        4.0      4.0      5.0  3.0    0.057325  
2        0.0      2.0      3.0  1.0    0.190476  
3        4.0      3.0      5.0  2.0    0.003841  
4        2.0      6.0      6.0  2.0    0.004854  
..       ...      ...      ...  ...         ...  
159      3.0      4.0      3.0  2.0    0.009360  
160      1.0      1.0      5.0  2.0    0.013294  
161      2.0      2.0      5.0  2.0    0.000000  
162      1.0      1.0      1.0  3.0    0.000000  
163      5.0      4.0      5.0  2.0    0.000000  

[160 rows x 10 columns]

In [3]:
full = df
%store full

Stored 'full' (DataFrame)


In [4]:
#Let's confirm that aspiration satisfy the relevance condition for performance distance to the best. There is a difference between aspiration and social aspiration. 
reg_expr = 'abs_perform_diff_best ~ Q10'

# Build and train an OLS model that regresses performance distance to the best on aspiration and verify
# using the F-test that coefficients of aspiration is significant 
olsr_model = smf.ols(formula=reg_expr, data=df)
olsr_model_results = olsr_model.fit()
print(olsr_model_results.summary())

df['ln_similarity'] = np.log(df['similarity'] + 1)

# Build out the exog matrix. Statsmodels requires this matrix to contain all the endogenous and
# exogenous variables, plus the constant.
exog = df[['abs_perform_diff_best']]
exog = add_constant(exog)

# Build out the instruments matrix. Statsmodels requires this matrix to contain not only all the
# instruments but also the variables in exog that will NOT be instrumented
instruments = df[['Q10']]
instruments = add_constant(instruments)

#Build and train the IV2SLS model
iv2sls_model = IV2SLS(endog=df['ln_similarity'], exog=exog, instrument=instruments)
iv2sls_model_results = iv2sls_model.fit()

#Print the training summary
print(iv2sls_model_results.summary())

#Compare the performance of 2SLS with OLS of ln(wage) on performance distance to the best
reg_expr = 'ln_similarity ~ abs_perform_diff_best'
olsr_model = smf.ols(formula=reg_expr, data=df)
olsr_model_results = olsr_model.fit()
print(olsr_model_results.summary())

                              OLS Regression Results                             
Dep. Variable:     abs_perform_diff_best   R-squared:                       0.086
Model:                               OLS   Adj. R-squared:                  0.080
Method:                    Least Squares   F-statistic:                     14.87
Date:                   Sat, 24 Sep 2022   Prob (F-statistic):           0.000167
Time:                           23:26:45   Log-Likelihood:                -913.78
No. Observations:                    160   AIC:                             1832.
Df Residuals:                        158   BIC:                             1838.
Df Model:                              1                                         
Covariance Type:               nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    258.9941 

In [5]:
## Phase 3

In [6]:
df = pd.read_csv('perform_dist_high_compare_b3.csv', header=0)
df = df.dropna()
df = df[df['phase'] == 3]
df

Unnamed: 0             Student  group  abs_perform_diff_best  phase  \
82           82           lemartinp      3                 193.33      3   
83           83           juligarji      3                 126.67      3   
84           84           nfmorenog      3                 196.67      3   
85           85           AFelipeGA      3                 177.50      3   
86           86          mballeng91      3                 230.00      3   
87           87            dduartec      3                  25.00      3   
88           88           srmedinac      3                 214.17      3   
89           89        wilson911013      3                 239.17      3   
90           90       wapiravaguens      3                 137.50      3   
91           91            davidh17      3                 109.17      3   
92           92      DianaNavarrete      3                 182.50      3   
93           93           jdmonroyg      3                 330.00      3   
94           94           jorge52an      3                 167.50      3   
95           95     jscastelblancoh      3                  99.17      3   
96           96          juclopezso      3                 218.33      3   
97           97           afforeroc      3                 226.67      3   
98           98         dagofonseca      3                 102.50      3   
99           99         lacardenasv      3                  65.83      3   
100         100         oacastillol      3                 209.17      3   
101         101          ncampuzano      3                 213.33      3   
102         102            adriel62      3                 139.17      3   
103         103            capinzor      3                 122.50      3   
104         104            ofnanezn      3                 181.67      3   
105         105           macuestap      3                 185.00      3   
106         106              spenas      3                 214.17      3   
107         107         kgbayala528      3                  78.33      3   
108         108          lgmoralesa      3                  20.00      3   
109         109          ronaldsg20      3                 172.50      3   
110         110       cccristanchoc      3                 226.67      3   
111         111          jmalvarezd      3                  55.00      3   
112         112           smarquezo      3                 218.33      3   
113         113             sarizag      3                 201.67      3   
114         114            Chranium      3                 235.00      3   
115         115             DavidQP      3                 205.83      3   
116         116         afceballosr      3                 226.67      3   
117         117           lizzyt10h      3                 184.17      3   
118         118          NicolasPrr      3                 221.67      3   
120         120          feartheGru      3                 330.00      3   
121         121        Danielsv9207      3                 330.00      3   
122         122  JhonEmmanuelTorres      3                 330.00      3   

     Q7_Q7_1  Q7_Q7_2  Q8_Q8_1  Q10  similarity  
82       0.0      0.0      4.0  1.0    0.100000  
83       4.0      4.0      5.0  3.0    0.009495  
84       0.0      2.0      3.0  1.0    0.139535  
85       4.0      3.0      5.0  2.0    0.010793  
86       2.0      6.0      6.0  2.0    0.007353  
87       1.0      5.0      4.0  1.0    0.092896  
88       1.0      2.0      4.0  1.0    0.055944  
89       0.0      1.0      2.0  1.0    0.010309  
90       0.0      1.0      4.0  2.0    0.071856  
91       3.0      3.0      5.0  3.0    0.164557  
92       0.0      1.0      5.0  2.0    0.030303  
93       0.0      1.0      4.0  1.0    0.000000  
94       4.0      4.0      5.0  2.0    0.044077  
95       0.0      0.0      3.0  2.0    0.083665  
96       1.0      3.0      3.0  2.0    0.059829  
97       1.0      4.0      3.0  2.0    0.051546  
98       0.0      0.0      2.0  2.0   

In [7]:
#Let's confirm that aspiration satisfy the relevance condition for performance distance to the best. There is a difference between aspiration and social aspiration. 
reg_expr = 'abs_perform_diff_best ~ Q10'

# Build and train an OLS model that regresses performance distance to the best on aspiration and verify
# using the F-test that coefficients of aspiration is significant 
olsr_model = smf.ols(formula=reg_expr, data=df)
olsr_model_results = olsr_model.fit()
print(olsr_model_results.summary())

df['ln_similarity'] = np.log(df['similarity'] + 1)

# Build out the exog matrix. Statsmodels requires this matrix to contain all the endogenous and
# exogenous variables, plus the constant.
exog = df[['abs_perform_diff_best']]
exog = add_constant(exog)

# Build out the instruments matrix. Statsmodels requires this matrix to contain not only all the
# instruments but also the variables in exog that will NOT be instrumented
instruments = df[['Q10']]
instruments = add_constant(instruments)

#Build and train the IV2SLS model
iv2sls_model = IV2SLS(endog=df['ln_similarity'], exog=exog, instrument=instruments)
iv2sls_model_results = iv2sls_model.fit()

#Print the training summary
print(iv2sls_model_results.summary())

#Compare the performance of 2SLS with OLS of ln(wage) on performance distance to the best
reg_expr = 'ln_similarity ~ abs_perform_diff_best'
olsr_model = smf.ols(formula=reg_expr, data=df)
olsr_model_results = olsr_model.fit()
print(olsr_model_results.summary())

                              OLS Regression Results                             
Dep. Variable:     abs_perform_diff_best   R-squared:                       0.076
Model:                               OLS   Adj. R-squared:                  0.052
Method:                    Least Squares   F-statistic:                     3.138
Date:                   Sat, 24 Sep 2022   Prob (F-statistic):             0.0845
Time:                           23:26:45   Log-Likelihood:                -228.46
No. Observations:                     40   AIC:                             460.9
Df Residuals:                         38   BIC:                             464.3
Df Model:                              1                                         
Covariance Type:               nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    236.2613 

## Phase 4

In [8]:
df = pd.read_csv('perform_dist_high_compare_b3.csv', header=0)
df = df.dropna()
df = df[df['phase'] == 4]
df

Unnamed: 0             Student  group  abs_perform_diff_best  phase  \
123         123           lemartinp      3                  57.50      4   
125         125           nfmorenog      3                 237.50      4   
126         126           AFelipeGA      3                 205.83      4   
127         127          mballeng91      3                 245.83      4   
128         128            dduartec      3                 270.83      4   
129         129           srmedinac      3                 242.50      4   
130         130        wilson911013      3                 275.83      4   
131         131       wapiravaguens      3                 140.83      4   
132         132            davidh17      3                 283.33      4   
133         133      DianaNavarrete      3                 242.50      4   
134         134           jdmonroyg      3                 358.33      4   
135         135           jorge52an      3                 115.00      4   
136         136     jscastelblancoh      3                  41.66      4   
137         137          juclopezso      3                 206.66      4   
138         138           afforeroc      3                 259.16      4   
139         139         dagofonseca      3                  93.33      4   
140         140         lacardenasv      3                  98.33      4   
141         141         oacastillol      3                 237.50      4   
142         142          ncampuzano      3                 213.33      4   
143         143            adriel62      3                 147.50      4   
144         144            capinzor      3                 163.33      4   
145         145            ofnanezn      3                 230.00      4   
146         146           macuestap      3                 209.16      4   
147         147              spenas      3                 242.50      4   
148         148         kgbayala528      3                 115.00      4   
149         149          lgmoralesa      3                  20.83      4   
150         150          ronaldsg20      3                 192.50      4   
151         151       cccristanchoc      3                 242.50      4   
152         152          jmalvarezd      3                 140.00      4   
153         153           smarquezo      3                 246.66      4   
154         154             sarizag      3                 167.50      4   
155         155            Chranium      3                 263.33      4   
156         156             DavidQP      3                  86.66      4   
157         157         afceballosr      3                  93.33      4   
158         158           lizzyt10h      3                 212.50      4   
159         159          NicolasPrr      3                 237.50      4   
160         160       jumcorredorro      3                  53.33      4   
161         161          feartheGru      3                 358.33      4   
162         162        Danielsv9207      3                 358.33      4   
163         163  JhonEmmanuelTorres      3                 358.33      4   

     Q7_Q7_1  Q7_Q7_2  Q8_Q8_1  Q10  similarity  
123      0.0      0.0      4.0  1.0    0.005437  
125      0.0      2.0      3.0  1.0    0.006931  
126      4.0      3.0      5.0  2.0    0.028254  
127      2.0      6.0      6.0  2.0    0.005132  
128      1.0      5.0      4.0  1.0    0.030488  
129      1.0      2.0      4.0  1.0    0.005760  
130      0.0      1.0      2.0  1.0    0.000390  
131      0.0      1.0      4.0  2.0    0.006834  
132      3.0      3.0      5.0  3.0    0.007602  
133      0.0      1.0      5.0  2.0    0.001559  
134      0.0      1.0      4.0  1.0    0.004287  
135      4.0      4.0      5.0  2.0    0.014080  
136      0.0      0.0      3.0  2.0    0.013668  
137      1.0      3.0      3.0  2.0    0.003871  
138      1.0      4.0      3.0  2.0    0.001949  
139      0.0      0.0      2.0  2.0    0.013228  
140      1.0      2.0      5.0  4.0   

In [9]:
#Let's confirm that aspiration satisfy the relevance condition for performance distance to the best. There is a difference between aspiration and social aspiration. 
reg_expr = 'abs_perform_diff_best ~ Q10'

# Build and train an OLS model that regresses performance distance to the best on aspiration and verify
# using the F-test that coefficients of aspiration is significant 
olsr_model = smf.ols(formula=reg_expr, data=df)
olsr_model_results = olsr_model.fit()
print(olsr_model_results.summary())

df['ln_similarity'] = np.log(df['similarity'] + 1)

# Build out the exog matrix. Statsmodels requires this matrix to contain all the endogenous and
# exogenous variables, plus the constant.
exog = df[['abs_perform_diff_best']]
exog = add_constant(exog)

# Build out the instruments matrix. Statsmodels requires this matrix to contain not only all the
# instruments but also the variables in exog that will NOT be instrumented
instruments = df[['Q10']]
instruments = add_constant(instruments)

#Build and train the IV2SLS model
iv2sls_model = IV2SLS(endog=df['ln_similarity'], exog=exog, instrument=instruments)
iv2sls_model_results = iv2sls_model.fit()

#Print the training summary
print(iv2sls_model_results.summary())

#Compare the performance of 2SLS with OLS of ln(wage) on performance distance to the best
reg_expr = 'ln_similarity ~ abs_perform_diff_best'
olsr_model = smf.ols(formula=reg_expr, data=df)
olsr_model_results = olsr_model.fit()
print(olsr_model_results.summary())

                              OLS Regression Results                             
Dep. Variable:     abs_perform_diff_best   R-squared:                       0.080
Model:                               OLS   Adj. R-squared:                  0.055
Method:                    Least Squares   F-statistic:                     3.290
Date:                   Sat, 24 Sep 2022   Prob (F-statistic):             0.0776
Time:                           23:26:45   Log-Likelihood:                -234.48
No. Observations:                     40   AIC:                             473.0
Df Residuals:                         38   BIC:                             476.3
Df Model:                              1                                         
Covariance Type:               nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    262.8576 

## Comments

![image1a](image1a.png)

The coefficients of aspiration (Q10) is significant at a p of < 0 as indicated by their p-values which are basically zero. Aspiration (Q10) clearly meet the **relevance condition** for instrumental variables of performance distance. 

We’ll now build a linear model for the similarity to the best equation and using statsmodels, we’ll train the model using the two-stage least square estimator.

We’ll start by building the design matrices. The dependent variable is ln(similarity). 

```
ln_wage = np.log(df['similarity'])
```

Statsmodel’s IV2SLS estimator is defined as follows:

```

statsmodels.sandbox.regression.gmm.IV2SLS(endog, exog, instrument=None)

```
Statsmodels needs the endog, exog and instrument matrices to be constructed in a specific way as follows:

`endog` is an [n x 1] matrix containing the dependent variable. In our example, it is the log(similarity) variable.

`exog` is an [n x (k+1)] size matrix that must contain all the endogenous and exogenous variables, plus the constant. In our example, apart from the constant, we do not have any exogenous variables defined in our wage equation. 

`instrument` is a matrix that contains the instrumental variables. Additionally, the Statsmodels’ IV2SLS estimator requires instrument to also contain all variables from the exog matrix that are not being instrumented. In our example, the instrumental variable is aspiration. The variables in exog that are not being instrumented is just the placeholder column for the intercept. 

![image2a](image2a.png)

## Interpretation of results of the 2SLS model

Since our primary interest is in estimating the effect of performance distance to the best on similarity to the best, we’ll focus our attention on the coefficient estimate of the performance distance to the best. 

We see that the 2SLS model has estimated the coefficient of performance distance to the best as -7.251e-05 with a standard error of 0.000 and a 95% confidence interval of -0.000 to 0.000. The p-value of 0.084 suggests a significance at (1–0.702)100%=29.8%. Overall, and as expected for a 2SLS model, the model lacks precision.

Note that dependent variable is log(similarity + 1). To calculate the rate of change of similarity to the best for each unit change of performance distance to the best, we must exponentiate the coefficient of performance distance to the best.

1 - 10^(-7.251e-05) = 0.000167 implying that a unit decrease in performance distance to the best is estimated to yield an increase of 0.000167 in similarity to the best, and vice-versa.

## Comparison of the IV estimator with an OLS estimator 

Let’s compare the performance of the 2SLS model with a straight-up OLS model that regresses log(similarity) on performance distance to the best. 

![image3a.png](image3a.png)

We’ll focus our attention on the estimated value of the coefficient of performance distance to the best. At -0.0002, it is a lot higher than the estimate reported by the 2SLS model.

1 - 10^(-0.0002)=0.00046, implying a unit decrease in the performance distance to the best is estimated to translate into a 0.00046 increase in similarity to the best (vice versa). 

The higher estimate from OLS is expected due to the suspected endogeniety of performance distance to the best. In practice, depending on the situation we are modeling, we may want to accept the more conservative estimate of -7.251e-05 reported by the 2SLS model. However, (and against the 2SLS model), the coefficient estimate from the OLS model is highly significant with a p-value that is 0.000. Recollect that the estimate from the 2SLS model was significant at only a 29.8% confidence level.

The coefficient estimate of performance distance to the best reported by the OLS model has pretty similar standard error which is zero as compared to that from the 2SLS model. 

For comparison, here are the coefficient estimates of performance distance to the best and corresponding 95% CIs from the two models:

With the IV estimator, one trades precision of estimates for the removal of endogeneity and the consequent bias in the estimates.

